gptme/server/tasks_api.py

```diff
@@ -0,0 +1,724 @@
+with workspace and git information derived from the active conversation.
+"""
+
+import json
+import logging
+import subprocess
+from dataclasses import asdict, dataclass, field
+from datetime import datetime
+from pathlib import Path
+from typing import Any, Literal
+
+import flask
+
+from ..dirs import get_logs_dir
+from ..logmanager import LogManager
+from ..config import ChatConfig
+from ..message import Message
+from ..prompts import get_prompt
+from ..tools import get_toolchain
+
+logger = logging.getLogger(__name__)
+
+tasks_api = flask.Blueprint("tasks_api", __name__)
+
+# Type definitions
+TaskStatus = Literal["pending", "active", "completed", "failed"]
+TargetType = Literal["stdout", "pr", "email", "tweet"]
+
+
+@dataclass
+class Task:
+    """Task metadata container."""
+
+    id: str
+    content: str
+    created_at: str
+    status: TaskStatus
+    target_type: TargetType
+    target_repo: str | None = None
+    conversation_ids: list[str] = field(default_factory=list)
+    metadata: dict[str, Any] = field(default_factory=dict)
+
+
+def get_tasks_dir() -> Path:
+    """Get the tasks storage directory."""
+    return get_logs_dir().parent / "tasks"
+
+
+def load_task(task_id: str) -> Task | None:
+    """Load a task from storage."""
+    tasks_dir = get_tasks_dir()
+    task_file = tasks_dir / f"{task_id}.json"
+
+    if not task_file.exists():
+        return None
+
+    try:
+        with open(task_file) as f:
+            data = json.load(f)
+        return Task(**data)
+    except Exception as e:
+        logger.error(f"Error loading task {task_id}: {e}")
+        return None
+
+
+def save_task(task: Task) -> None:
+    """Save a task to storage."""
+    tasks_dir = get_tasks_dir()
+    tasks_dir.mkdir(parents=True, exist_ok=True)
+
+    task_file = tasks_dir / f"{task.id}.json"
+
+    try:
+        with open(task_file, "w") as f:
+            json.dump(asdict(task), f, indent=2)
+    except Exception as e:
+        logger.error(f"Error saving task {task.id}: {e}")
+        raise
+
+
+def delete_task_file(task_id: str) -> bool:
+    """Delete a task file from storage."""
+    tasks_dir = get_tasks_dir()
+    task_file = tasks_dir / f"{task_id}.json"
+
+    if task_file.exists():
+        try:
+            task_file.unlink()
+            return True
+        except Exception as e:
+            logger.error(f"Error deleting task {task_id}: {e}")
+            return False
+    return False
+
+
+def list_tasks() -> list[Task]:
+    """List all tasks from storage."""
+    tasks_dir = get_tasks_dir()
+
+    if not tasks_dir.exists():
+        return []
+
+    tasks = []
+    for task_file in tasks_dir.glob("*.json"):
+        task_id = task_file.stem
+        task = load_task(task_id)
+        if task:
+            tasks.append(task)
+
+    return sorted(tasks, key=lambda t: t.created_at, reverse=True)
+
+
+def derive_task_status(task: Task) -> TaskStatus:
+    """Derive task status from conversation state and process information."""
```

Comment:
Poor status check. The web UI handles things like git status, PR state, etc. (takes a little while)
Maybe we should just list tasks with cached/stored status, and then on actual task retrieval we run an improved `derive_task_status` (maybe not for already done tasks) and persist the new state before returning it via API.